api/lib/opentelemetry/distributed_context/entry.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module DistributedContext
+    # An Entry consists of Entry::Metadata, Entry::Key, and Entry::Value.
+    class Entry
+      attr_reader :metadata, :key, :value
+
+      # Entry::Key is the name of the Entry. Entry::Key along with Entry::Value can be used to aggregate and group stats,
+      # annotate traces and logs, etc.
+      #
+      # Restrictions
+      # - Must contain only printable ASCII (codes between 32 and 126 inclusive)
+      # - Must have length greater than zero and less than 256.
+      # - Must not be empty.
+      class Key
+        attr_reader :name
+
+        def initialize(name)
+          # TODO: name should be printable ascii (32..126), not just ascii.
+          raise ArgumentError unless name&.ascii_compatible? && (1..255).include?(name.length)
+
+          @name = -name
+        end
+      end
+
+      # Entry::Value wraps a string. It MUST contain only printable ASCII (codes between 32 and 126).
+      class Value
+        def initialize(value)
+          # TODO: value should be printable ascii (32..126), not just ascii.
+          raise ArgumentError unless value&.ascii_compatible?
+
+          @value = -value
+        end
+
+        def to_s
+          @value
+        end
+      end
+
+      # Entry::Metadata contains properties associated with an Entry. For now only the property entry_ttl is defined.
+      # In future, additional properties may be added to address specific situations.
+      #
+      # The creator of entries determines metadata of an entry it creates.
+      class Metadata
+        attr_reader :entry_ttl
+
+        # An @see Entry with NO_PROPAGATION is considered to have local scope and is used within the process
+        # where it is created.
+        NO_PROPAGATION = 0
+
+        # An @see Entry with UNLIMITED_PROPAGATION can propagate unlimited hops. However, it is still subject
+        # to outgoing and incoming (on remote side) filter criteria.
+        UNLIMITED_PROPAGATION = -1
+
+        def initialize(entry_ttl)
+          raise ArgumentError unless entry_ttl&.integer?
```

Comment:
If we only want to allow integers and disallow strings for example then `unless entry_ttl.is_a?(Integer)` should work better. `#integer?` is not defined in non numeric classes.